src/lib.rs

```diff
@@ -677,7 +656,77 @@ pub unsafe extern "C" fn fd_readdir(
+    return State::with(|state| {
+        let dir = state.get_dir(fd)?;
+        let stream = wasi_filesystem::readdir(dir.fd)?;
+        // Schedule `stream` to get closed when the loop is finished below
+        let _dtor = CloseOnDrop(stream);
+
+        let mut i = 0;
+        while buf.len() > 0 {
+            i += 1;
+            // One allocation will be performed by the canonical ABI to make
+            // room for the pathname being read. Store it in the state's
+            // internal buffer which we'll copy out of later. Technically this
+            // could use `buf` if `buf` is bigger than `PATH_MAX +
+            // size_of::<Dirent>()` but that degree of optimization is left for
+            // a future pass.
+            state.register_buffer(state.path_buf.get().cast(), PATH_MAX);
+            let entry = match wasi_filesystem::read_dir_entry(stream)? {
+                Some(entry) => entry,
+                None => break,
+            };
+
+            let wasi_filesystem::DirEntry { ino, type_, name } = entry;
+            // Cancel the destructor of `name` which is a `String` since it
+            // lives in the memory within `state`.
+            let name = ManuallyDrop::new(name);
+
+            // The `cookie` is intepreted as the nth iteration of the stream
+            // that we're given. The `i` counter points one past the current
+            // iteration so skip this entry if the cookie is too large.
+            if cookie >= i {
+                continue;
+            }
+
+            // Copy a `dirent` describing this entry into the destination `buf`,
+            // truncating it if it doesn't fit entirely.
+            let dirent = wasi::Dirent {
+                d_next: i,
+                d_ino: ino.unwrap_or(0),
+                d_namlen: u32::try_from(name.len()).unwrap(),
+                d_type: type_.into(),
+            };
+            let bytes = core::slice::from_raw_parts(
+                (&dirent as *const wasi::Dirent).cast::<u8>(),
+                size_of::<Dirent>(),
+            );
+            let dirent_bytes_to_copy = buf.len().min(bytes.len());
+            buf[..dirent_bytes_to_copy].copy_from_slice(&bytes[..dirent_bytes_to_copy]);
+            buf = &mut buf[dirent_bytes_to_copy..];
+
+            // Copy the name bytes into the output `buf`, truncating it if it
+            // doesn't fit.
+            //
+            // Note that this might be a 0-byte copy if the `dirent` was
+            // truncated or fit entirely into the destination.
+            let name_bytes_to_copy = buf.len().min(name.len());
+            buf[..name_bytes_to_copy].copy_from_slice(&name[..name_bytes_to_copy]);
+            buf = &mut buf[name_bytes_to_copy..];
+        }
+
+        *bufused = buf_len - buf.len();
+        Ok(())
+    });
+
+    struct CloseOnDrop(wasi_filesystem::DirEntryStream);
+
+    impl Drop for CloseOnDrop {
+        fn drop(&mut self) {
+            wasi_filesystem::close(self.0);
```

Comment:
Also worth asking: should I add a separate `close` function for `dir-entry-stream`?

Reply:
Yes, also an oversight.